cilium, socklb: Terminate both UDP and TCP sockets

borkmann opened the pull request:
(see commits)

borkmann committed:
cilium, socklb: Terminate both UDP and TCP sockets

Extend the socket termination logic and also terminate ongoing TCP
connections when a backend goes down which allows for short feedback
time since it does not require traffic to be radiated from clients
to receive a RST from the remote.

Note in terminateConnectionsToBackend we also remove the lb.ANY as
they are removed by the reconciler in the new control plane once the
maps have been populated with the new entries.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Link: https://github.com/cilium/cilium/pull/39800/files#r2122023563

borkmann committed:
cilium, bpf: Add comment on revnat map wrt L4 protos

Add a small comment explaining why we don't have the L4 proto as part
of the lookup key.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ysksuzuki commented on pkg/loadbalancer/reconciler/termination.go:
```diff
@@ -180,17 +182,22 @@ func terminateUDPConnectionsToBackend(p socketTerminationParams, l3n4Addr lb.L3n
 	l4Addr := l3n4Addr.L4Addr
 
 	switch l3n4Addr.Protocol {
-	case lb.UDP, lb.ANY:
```
Do we still have the flag to disable the protocol differentiation? I thought when it was turned off, everything came in as ANY.

borkmann replied:
I've added a comment to the commit msg on why its ok to remove.

ysksuzuki commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -88,18 +88,18 @@ func Destroy(logger *slog.Logger, filter SocketFilter) error {
 				logger.Info("", logfields.Socket, sock)
-				if err := destroySocket(logger, sock, family, unix.IPPROTO_UDP, 0xffff, true); err != nil {
-					errs = errors.Join(errs, fmt.Errorf("destroying UDP socket with filter [%v]: %w", filter, err))
+				if err := destroySocket(logger, sock, family, protocol, 0xffff, true); err != nil {
```
If we set stateFilter to 0xffff, wouldn't that also remove sockets in the TCP_TIME_WAIT state, making the source address and port immediately reusable?

borkmann replied:
Hm, but this is for a backend which is gone, so if a new connection is created with the same tuples from a client, then a new backend is going to get picked, no?

borkmann replied:
Also, in TCP_TIME_WAIT we would have already removed the sock from the revnat map, see the removal at the release https://github.com/cilium/cilium/blob/main/bpf/bpf_sock.c#L1250 hook. The side that sends the first FIN (app triggering close) and ends up sending the last ACK goes into TIME_WAIT. Because it’s possible that the final ACK is lost, and the peer may retransmit the FIN. The TIME_WAIT state ensures that the socket remains around long enough to retransmit the ACK if needed, and to prevent delayed segments from contaminating a new connection. Given the client triggered the close in that case, we removed the client socket with its cookie from the map, and therefore it won't be found anymore from destroySocket.

borkmann committed:
cilium, socklb: Add states filter for the termination iteration

Add a state filter to the iterator and skip TCP sockets which are in
closing or time wait state. There is no need to spend time to iterate
these. Technically, there is no harm since when the client app closes
the socket and it goes into time wait state, then upon close the socket
LB removes the socket from the revnat map in cil_sock_release.. but
then again, no need to iterate through these.

Suggested-by: Yusuke Suzuki <yusuke.suzuki@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ysksuzuki reviewed (approved):
LGTM. Thanks!

julianwiedmann commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -31,6 +31,53 @@ var (
+	netlink.TCP_NEW_SYN_REC,
+	// Sockets in TCP_LISTEN are moved to closing state
+	// (net/ipv4/tcp.c#L4908)
+	netlink.TCP_CLOSE,
```
@borkmann I'm a bit confused here - in the code comment above we say that we are omitting some states in the filter (including `TCP_CLOSE`). But then it shows up in the filter after all.

julianwiedmann commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -31,6 +31,53 @@ var (
+	netlink.TCP_SYN_SENT,
+	netlink.TCP_CLOSING,
+	netlink.TCP_LAST_ACK,
+	netlink.TCP_LISTEN,
```
@borkmann @ysksuzuki given that we're only interested in terminating clients, can/should we drop `TCP_LISTEN` from the filter?
